src/utils/navigate.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * Helper for calling navigate that will append search query to path,
+ * if present, before appending an optional hash.
+ *
+ * Either search or hash may be passed as null|undefined, but if they
+ * are present must either be the empty string or start with their
+ * standard delimeter.
+ *
+ * @param {Function} navigate
+ * @param {string} path
+ * @param {object} options
+ */
+export function navWith(navigate, path, options = {
+  search: '',
+  hash: '',
+}) {
+  const search = options.search || ''
```

Comment:
I'd prefer the default behaviour to be maintaining current parameters,
```
export function navWith(navigate, path, options = {
  search: location.search,
  hash: location.hash,
}) {
```
 that way you wouldn't need the function `navWithSearchAndHashState` as well, if someone wants to preserve the state (query and hash) he can call navWith with 2 parameters, and if he wants to override any of which he should just pass the options parameters similar to what you did here 

``` 
 navWith(navigate, defaultPath, {
    search: location.search,
    hash: cameraHash,
  })
  ```

What do you think?

Reply:
Done.